### PR TITLE
Fix reporter bug when rejecting non-errors

### PR DESCRIPTION
**Pull request by jakwuh**
To reproduce try ava with the following test:

```js
import test from 'ava';

test(t => {
    Promise.reject(1);
});
```

### CODE REVIEW

**Comment by novemberborn** [comment]
Interesting find. I think the error comes from this line, really: https://github.com/avajs/ava/blob/98dded553c8c213739cea9743ab4662f40d3125a/lib/run-status.js#L24

Unhandled rejections are normalized: https://github.com/avajs/ava/blob/98dded553c8c213739cea9743ab4662f40d3125a/lib/run-status.js#L72

Perhaps the `message` should be set to `String(err)` and `stack` shouldn't be set at all. Further, error objects that have a non-string `message`, `name` or `stack` should have those properties deleted. That would be in https://github.com/avajs/ava/blob/98dded553c8c213739cea9743ab4662f40d3125a/lib/serialize-error.js#L59.

Long story short, this means errors have the expected shape in the reporters and we don't have to keep checking if certain properties are strings etc. @avajs/core thoughts?

**Comment by sindresorhus** [comment]
> Long story short, this means errors have the expected shape in the reporters and we don't have to keep checking if certain properties are strings etc. @avajs/core thoughts?

👍

**Comment by jakwuh** [comment]
@novemberborn thank you for the response. I think having `message: String(err)` + removing `stack` property is the best solution here. Isn't it?

**Comment by novemberborn** [comment]
@jakwuh yea. Though that's just for rejection reasons that aren't objects. We also need to sanitize objects with unexpected properties.

**Comment by jakwuh** [comment]
@novemberborn here we go, now regardfully listening to any feedback :)

File: lib/run-status.js [hunk 1]
```diff
@@ -18,10 +18,15 @@ function sum(arr, key) {
 }
 
 function normalizeError(err) {
-	if (!isObj(err)) {
+	if (isObj(err) && 'message' in err && typeof err.message === 'string') {
+		['name', 'stack'].forEach(field => {
+			if (field in err && typeof err[field] !== 'string') {
+				delete err[field];
```
Comment (novemberborn) [inline lib/run-status.js]: I think this really does need to go here: https://github.com/avajs/ava/blob/98dded553c8c213739cea9743ab4662f40d3125a/lib/serialize-error.js#L59

We don't want to modify `err` since it comes from user code. There may be side-effects. But we can't easily copy error objects since `name` and `message` aren't always enumerable. After the call to `cleanYamlObject` in `serialize-error.js` we have a copy we can safely modify.

Also, in this case, I'd prefer a little bit of duplication over your `forEach` trick. IMHO this is a lot easier to follow:

```js
if ('name' in err && typeof err.name !== 'string') {
  delete err.name
}
if ('stack' in err && typeof err.stack !== 'string') {
  delete err.stack
}
```

Even better, `delete` only deletes own-properties, so the `'name' in err` guard is unnecessary.

Comment (jakwuh) [inline lib/run-status.js]: Yes, this totally makes sense, still I find fairly to say that errors coming to `normalizeError` are already serialized (as I can see from the [test-worker](https://github.com/avajs/ava/blob/master/lib/test-worker.js#L42) code).
Also thank you a lot for your fast and detailed feedback.

**Comment by jakwuh** [comment]
@novemberborn here we go 🙂

File: lib/serialize-error.js [hunk 2]
```diff
@@ -20,6 +20,11 @@ function filter(propertyName, isRoot, source, target) {
 		return true;
 	}
 
+	if ((propertyName === 'name' || propertyName === 'stack') &&
+		typeof source[propertyName] !== 'string') {
```
Comment (novemberborn) [inline lib/serialize-error.js]: @sindresorhus any thoughts on this? XO doesn't complain 😉

Comment (sindresorhus) [inline lib/serialize-error.js]: Why should it have complained?

**Comment by sindresorhus** [comment]
Woot! Very good work @jakwuh :)

File: lib/serialize-error.js (hunk 2 above)
Comment (novemberborn) [inline lib/serialize-error.js]: Figured it may have opinions on if condition indenting.

**Comment by jakwuh** [comment]
Thanks  😊
